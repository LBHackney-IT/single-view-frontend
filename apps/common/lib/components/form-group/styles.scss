@import "../../../node_modules/lbh-frontend/lbh/base";
@import "../../../node_modules/govuk-frontend/govuk/objects/form-group";
@import "../../../node_modules/lbh-frontend/lbh/objects/form-group";
@import "../../../node_modules/lbh-frontend/lbh/components/lbh-label/label";
@import "../../../node_modules/lbh-frontend/lbh/components/lbh-hint/hint";
@import "../../../node_modules/lbh-frontend/lbh/components/lbh-error-message/error-message";

.lbh-form-group {
  label,
  legend {
    sup {
      color: lbh-colour("lbh-error");
    }
  }
}

fieldset.lbh-form-group legend {
  display: contents;
  margin-bottom: 0.375rem;
}
