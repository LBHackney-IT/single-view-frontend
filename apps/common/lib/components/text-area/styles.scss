@import "../../../node_modules/lbh-frontend/lbh/base";
@import "../../../node_modules/lbh-frontend/lbh/components/lbh-textarea/textarea";
@import "../../../node_modules/lbh-frontend/lbh/components/lbh-character-count/character-count";
