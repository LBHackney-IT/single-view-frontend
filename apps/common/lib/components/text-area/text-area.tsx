import React, {
  ChangeEvent,
  ComponentPropsWithoutRef,
  forwardRef,
  useCallback,
  useMemo,
  useState,
} from "react";

import classNames from "classnames";

import { pluralize, widthOverrides } from "../../utils";

import "./styles.scss";

export interface TextAreaProps extends ComponentPropsWithoutRef<"textarea"> {
  maxLength?: number;
  error?: boolean;
  override?: number;
}

const getLengthOfValue = (
  initialValue: string | number | readonly string[] | undefined,
) => {
  if (typeof initialValue === "string") {
    return initialValue.length;
  }
  if (Array.isArray(initialValue)) {
    return initialValue.join(",").length;
  }
  return String(initialValue || "").length;
};

export const TextArea = forwardRef<HTMLTextAreaElement, TextAreaProps>(function TextArea(
  { maxLength, error, className, onChange, override, ...props },
  ref,
) {
  const { value, defaultValue } = props;
  const isControlled = value !== undefined;
  const initialValue = value || defaultValue;

  const [characterCount, setCharacterCount] = useState(getLengthOfValue(initialValue));

  const onChangeHandler = useCallback(
    (event: ChangeEvent<HTMLTextAreaElement>) => {
      if (
        event?.currentTarget?.value !== undefined &&
        !isControlled &&
        maxLength !== undefined
      ) {
        setCharacterCount(String(event.currentTarget.value).length);
      }

      if (typeof onChange === "function") {
        onChange(event);
      }
    },
    [onChange, maxLength, isControlled],
  );

  const exceedingValue = useMemo(
    () =>
      maxLength !== undefined &&
      maxLength - (isControlled ? getLengthOfValue(value) : characterCount),
    [maxLength, characterCount, value, isControlled],
  );

  const textareaClasses = classNames(
    "govuk-textarea",
    "lbh-textarea",
    { "govuk-textarea--error": error },
    "lbh-character-count",
    widthOverrides(override),
    className,
  );

  const messageClasses = classNames(
    { "govuk-hint": exceedingValue !== false && exceedingValue >= 0 },
    "govuk-character-count__message",
    {
      "govuk-error-message": exceedingValue !== false && exceedingValue < 0,
    },
    widthOverrides(override),
  );

  return (
    <>
      <textarea
        ref={ref}
        className={textareaClasses}
        onChange={onChangeHandler}
        {...props}
      />
      {maxLength !== undefined && exceedingValue !== false && (
        <span className={messageClasses} aria-live="polite">
          {exceedingValue >= 0
            ? `You have ${exceedingValue} ${pluralize(
                "character",
                exceedingValue,
              )} remaining`
            : `You have ${Math.abs(exceedingValue)} ${pluralize(
                "character",
                exceedingValue,
              )} too many`}
        </span>
      )}
    </>
  );
});
