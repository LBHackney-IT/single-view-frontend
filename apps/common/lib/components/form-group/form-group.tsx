import React, {
  Children,
  ReactElement,
  cloneElement,
  forwardRef,
  isValidElement,
  useMemo,
} from "react";

import classNames from "classnames";

import { widthOverrides } from "../../utils";
import { TextArea } from "../text-area";

import type * as Polymorphic from "@radix-ui/react-polymorphic";
import "./styles.scss";

export interface FormGroupProps {
  id: string;
  label?: string;
  name?: string;
  hint?: string;
  error?: string | false;
  required?: boolean;
  children: ReactElement;
  override?: number;
}

export type FormGroupComponent = Polymorphic.ForwardRefComponent<"div", FormGroupProps>;

export const FormGroup: FormGroupComponent = forwardRef(function FormGroup(
  {
    as: FormGroupComp = "div",
    id,
    name,
    label,
    hint,
    error,
    required,
    children,
    className,
    override,
    ...props
  },
  ref,
) {
  const formGroupClasses = classNames(
    "govuk-form-group",
    {
      "govuk-form-group--error": !!error,
    },
    "lbh-form-group",
    widthOverrides(override),
    className,
  );

  const describedBy = useMemo(() => {
    const classes: string[] = [];
    if (hint) {
      classes.push(`${id}-hint`);
    }
    if (error) {
      classes.push(`${id}-error`);
    }
    return classes.join(" ");
  }, [id, hint, error]);

  const LabelComp =
    typeof FormGroupComp === "string" && (FormGroupComp as string) === "fieldset"
      ? "legend"
      : "label";

  const formGroup = (
    <FormGroupComp ref={ref} id={id} className={formGroupClasses} {...props}>
      <LabelComp className="govuk-label lbh-label" htmlFor={`${id}-field`}>
        {label}
        {required ? <sup aria-hidden="true">*</sup> : ""}
      </LabelComp>
      {!!hint && (
        <span id={`${id}-hint`} className="govuk-hint lbh-hint">
          {hint}
        </span>
      )}
      {!!error && (
        <span id={`${id}-error`} className="govuk-error-message lbh-error-message">
          <span className="govuk-visually-hidden">Error:</span> {error}
        </span>
      )}
      {!!children &&
        Children.only(
          cloneElement(children, {
            id: `${id}-field`,
            name,
            required,
            error: !!error,
            "aria-describedby": describedBy || undefined,
            ...children.props,
          }),
        )}
    </FormGroupComp>
  );

  return isValidElement(children) && children.type === TextArea ? (
    <div className="govuk-character-count">{formGroup}</div>
  ) : (
    formGroup
  );
});
