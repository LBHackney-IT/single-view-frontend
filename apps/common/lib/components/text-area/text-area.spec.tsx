import React, { ChangeEvent, useState } from "react";

import { render, testA11y } from "@hackney/mtfh-test-utils";
import { fireEvent, screen } from "@testing-library/react";

import { FormGroup } from "../form-group";
import { TextArea } from "./text-area";

test("it passes a11y", async () => {
  const { container } = render(
    <FormGroup id="test" name="test" label="TextArea Test">
      <TextArea />
    </FormGroup>,
  );
  expect(container).toMatchSnapshot();
  await testA11y(container);
});

test("it renders correctly", () => {
  const { container } = render(<TextArea />);
  expect(container).toMatchSnapshot();
});

test("it renders with character count if maxLength is set", () => {
  render(<TextArea maxLength={500} />);
  expect(screen.getByText("You have 500 characters remaining")).toBeInTheDocument();
});

test("it renders with character count with correct plural", () => {
  render(<TextArea maxLength={2} placeholder="TextArea" />);
  const textArea = screen.getByPlaceholderText("TextArea");

  expect(screen.getByText("You have 2 characters remaining")).toBeInTheDocument();
  fireEvent.change(textArea, { target: { value: "A" } });
  expect(screen.getByText("You have 1 character remaining")).toBeInTheDocument();
  fireEvent.change(textArea, { target: { value: "AB" } });
  expect(screen.getByText("You have 0 characters remaining")).toBeInTheDocument();
  fireEvent.change(textArea, { target: { value: "ABC" } });
  expect(screen.getByText("You have 1 character too many")).toBeInTheDocument();
  fireEvent.change(textArea, { target: { value: "ABCD" } });
  expect(screen.getByText("You have 2 characters too many")).toBeInTheDocument();
});

test("it decreases the character count when text area value increases", () => {
  render(<TextArea maxLength={500} placeholder="TextArea" />);
  const textArea = screen.getByPlaceholderText("TextArea");
  fireEvent.change(textArea, { target: { value: "Some text" } });
  expect(screen.getByText("You have 491 characters remaining")).toBeInTheDocument();
});

test("it renders with character count if component is controlled", () => {
  render(<TextArea maxLength={500} value="Some text" />);
  expect(screen.getByText("You have 491 characters remaining")).toBeInTheDocument();
});

test("it renders with character count with array value", () => {
  render(<TextArea maxLength={500} defaultValue={["Some", "text"]} />);
  expect(screen.getByText("You have 491 characters remaining")).toBeInTheDocument();
});

test("it renders with character count with number value", () => {
  render(<TextArea maxLength={5} defaultValue={1234} />);
  expect(screen.getByText("You have 1 character remaining")).toBeInTheDocument();
});

test("it renders with character count with undefined value", () => {
  render(<TextArea maxLength={5} defaultValue={undefined} />);
  expect(screen.getByText("You have 5 characters remaining")).toBeInTheDocument();
});

test("it supports an onChange handler", () => {
  const onChange = jest.fn<void, ChangeEvent<HTMLTextAreaElement>[]>();
  render(<TextArea placeholder="TextArea" onChange={onChange} />);
  const textArea = screen.getByPlaceholderText("TextArea");
  fireEvent.change(textArea, { target: { value: "Some text" } });
  expect(onChange).toHaveBeenCalledTimes(1);
  expect(onChange.mock.calls[0][0].target.value).toContain("Some text");
});

test("it supports an onChange handler with maxLength", () => {
  const onChange = jest.fn<void, ChangeEvent<HTMLTextAreaElement>[]>();
  render(<TextArea maxLength={500} placeholder="TextArea" onChange={onChange} />);
  const textArea = screen.getByPlaceholderText("TextArea");
  fireEvent.change(textArea, { target: { value: "Some text" } });
  expect(onChange).toHaveBeenCalledTimes(1);
  expect(onChange.mock.calls[0][0].target.value).toContain("Some text");
});

test("it supports a controlled component", () => {
  const ControlledComponentTest = () => {
    const [value, setValue] = useState("");
    return (
      <>
        <TextArea
          maxLength={10}
          value={value}
          onChange={(event: ChangeEvent<HTMLTextAreaElement>) =>
            setValue(event.currentTarget.value)
          }
          placeholder="TextArea"
        />
        <button type="button" onClick={() => setValue("Hello")}>
          Test button
        </button>
      </>
    );
  };
  render(<ControlledComponentTest />);
  const textArea = screen.getByPlaceholderText("TextArea") as HTMLTextAreaElement;
  fireEvent.change(textArea, { target: { value: "Some text" } });
  expect(textArea.value).toBe("Some text");
  const button = screen.getByText("Test button");
  fireEvent.click(button);
  expect(textArea.value).toBe("Hello");
  expect(screen.getByText("You have 5 characters remaining")).toBeInTheDocument();
});
