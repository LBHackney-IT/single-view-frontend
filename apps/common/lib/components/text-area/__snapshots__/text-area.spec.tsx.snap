// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`it passes a11y 1`] = `
<div>
  <div
    class="govuk-character-count"
  >
    <div
      class="govuk-form-group lbh-form-group"
      id="test"
    >
      <label
        class="govuk-label lbh-label"
        for="test-field"
      >
        TextArea Test
        
      </label>
      <textarea
        class="govuk-textarea lbh-textarea lbh-character-count"
        id="test-field"
        name="test"
      />
    </div>
  </div>
</div>
`;

exports[`it renders correctly 1`] = `
<div>
  <textarea
    class="govuk-textarea lbh-textarea lbh-character-count"
  />
</div>
`;
